src/layer4/vxlan.rs

```diff
@@ -0,0 +1,166 @@
+use crate::{
+    errors::Error,
+    record::{
+        PcapRecord
+    },
+    layer4::udp::Udp,
+    layer2::ethernet::Ethernet,
+};
+
+use nom::{
+    *,
+    Err as NomError,
+    ErrorKind as NomErrorKind,
+    IResult
+};
+
+pub enum MaybeVxlan<'a> {
+    Encapsulated(Vxlan<'a>),
+    NotEncapsulated(&'a Udp<'a>)
+}
+
+#[derive(Debug)]
+pub struct Vxlan<'a> {
+    flags: u16,
+    group_policy_id: u16,
+    network_identifier: u32, // only can use 3 bytes
+    payload: &'a [u8],
+}
+
+impl<'a> Vxlan<'a> {
+    pub fn flags(&self) -> u16 {
+        self.flags
+    }
+    pub fn group_policy_id(&self) -> u16 {
+        self.group_policy_id
+    }
+    /// Not lager than 2^24
+    pub fn network_identifier(&self) -> u32 {
+        self.network_identifier
+    }
+    pub fn payload(&self) -> &[u8] {
+        self.payload
+    }
+
+    pub fn new(
+        flags: u16,
+        group_policy_id: u16,
+        network_identifier: u32,
+        payload: &[u8]
+    ) -> Vxlan {
+        Vxlan {
+            flags,
+            group_policy_id,
+            network_identifier,
+            payload,
+        }
+    }
+
+    pub fn parse<'b>(input: &'b Udp<'b>, endianness: nom::Endianness) -> Result<Vxlan<'b>, Error> {
+        /// TODO: Is Endianness really unknown?
+        do_parse!(input.payload(),
+            flags: u16!(endianness) >>
+            group_policy_id: u16!(endianness) >>
+            network_identifier: u32!(endianness) >> // actually u24 plus 8 reserved bits.
+            //reserved: be_u8 >> // accounted for in bytes captured under network_identifier
+            payload: rest >> (
+                (flags, group_policy_id, network_identifier, payload)
+            )
+        ).map_err(Error::from)
+            .and_then(|(_, parsed)| {
+                Ok(Vxlan {
+                    flags: parsed.0,
+                    group_policy_id: parsed.1,
+                    network_identifier: parsed.2>>8,
+                    payload: parsed.3
+                })
+            })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        layer2::ethernet::Ethernet,
+        layer3::ipv4::IPv4,
+        layer4::{
+            udp::Udp,
+            vxlan::Vxlan,
+        },
+        tests::util::parse_hex_dump,
+    };
+
+    #[test]
+    fn encapsulated() {
+        let bytes = parse_hex_dump(r##"
+            # Frame 3: 148 bytes on wire (1184 bits), 148 bytes captured (1184 bits) on interface 0
```

Comment:
This looks like a dump, is it from your computer or some commonly available pcap?